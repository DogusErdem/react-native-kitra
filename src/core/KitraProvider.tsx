import type { ComponentType, ForwardedRef } from 'react';
import React from 'react';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { StyleProp, ViewStyle } from 'react-native';

import { ThemeProvider } from './theme/theme';
import { TypographyContext, TypographyProvider } from './typography/typography';
import { NotificationContextType, NotificationProvider } from './notification/notification';

const Feather = require('react-native-vector-icons/Feather').default;
const Ionicons = require('react-native-vector-icons/Ionicons').default;
const AntDesign = require('react-native-vector-icons/AntDesign').default;

Animated.addWhitelistedNativeProps({ text: true });
const messageTypes = (theme:any) => ({
  SUCCESS: {
    backgroundColor: theme.colors.status.successLight,
    icon: <Feather name="check" size={25} color={theme.colors.status.success} />,
  },
  INFO: {
    backgroundColor: theme.colors.system.primary15,
    icon: <Ionicons name="information-circle-outline" size={25} color={theme.colors.system.primary} />,
  },
  DANGER: {
    backgroundColor: theme.colors.status.errorLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.error} />,
  },
  WARNING: {
    backgroundColor: theme.colors.status.warningLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.warning} />,
  },
});

type KitraProviderType= {
  children?:any,
  limit?:number,
  notificationRef?:React.Ref<NotificationContextType>,
  messageType?:(theme?:any, colorScheme?:'dark' | 'light') => {[key:string]: {
      backgroundColor:string,
    icon?:React.ReactNode
    onPress?:()=>void
  }},
  notificationCcontainerStyle?:StyleProp<ViewStyle>,
  customView?:({ header, type, message, theme }:
    {header:string, type:string, message:string, theme:any})=>React.ReactNode}

export const KitraProvider =
({ children, messageType = messageTypes, notificationCcontainerStyle, limit, customView,
  notificationRef }: KitraProviderType) => (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <ThemeProvider>
        <TypographyProvider>
          <NotificationProvider
            messageType={theme => (messageType ? messageType(theme) : theme)}
            notificationCcontainerStyle={notificationCcontainerStyle}
            limit={limit}
            customView={item => customView?.(item)}
            ref={notificationRef}
          >
            {children}
          </NotificationProvider>
        </TypographyProvider>
      </ThemeProvider>
    </GestureHandlerRootView>
);

export function applyDefaults<T>(Component: ComponentType<T>) {
  return React.forwardRef((props: T, ref: ForwardedRef<any>) => (
    <TypographyContext.Consumer>
      {typography =>
        <Component ref={ref} typography={typography?.typography} {...props} />
          }
    </TypographyContext.Consumer>
  ));
}
