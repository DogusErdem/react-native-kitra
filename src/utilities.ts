/* eslint-disable no-bitwise */

import type React from 'react';
import type { IconType } from './types';
import { x11Colors } from './core/theme/x11';

function hexToRgbA(hex:string, percentage:number) {
  if (hex[0] !== '#') {
    hex = `#${x11ColorToHex(hex)}`;
  }

  let c;
  if (/^#([A-Fa-f0-9]{3}){1,2}$/.test(hex)) {
    c = hex.substring(1).split('');
    if (c.length === 3) {
      c = [c[0], c[0], c[1], c[1], c[2], c[2]];
    }
    c = `0x${c.join('')}`;
    // @ts-ignore
    return `rgba(${[(c >> 16) & 255, (c >> 8) & 255, c & 255].join(',')},${percentage / 100})`;
  }
  throw new Error('Bad Hex');
}

export const opacity = (color:string, percentage: number) => hexToRgbA(color, percentage);

export function x11ColorToHex(colorName: string) {
  const hex = x11Colors.find(x => x.name === colorName)?.hex;
  if (!hex) {
    throw new Error(`Unknown X11 color name: ${colorName}`);
  }
  return hex;
}

const customIcons: any = {};

export const registerCustomIconType = (id: string, customIcon: any) => {
  customIcons[id] = customIcon;
};

export const getIconType = (type: IconType): any => {
  switch (type) {
    case 'zocial':
      return require('react-native-vector-icons/Zocial').default;
    case 'octicon':
      return require('react-native-vector-icons/Octicons').default;
    case 'material':
      return require('react-native-vector-icons/MaterialIcons').default;
    case 'material-community':
      return require('react-native-vector-icons/MaterialCommunityIcons')
        .default;
    case 'ionicon':
      return require('react-native-vector-icons/Ionicons').default;
    case 'foundation':
      return require('react-native-vector-icons/Foundation').default;
    case 'evilicon':
      return require('react-native-vector-icons/EvilIcons').default;
    case 'entypo':
      return require('react-native-vector-icons/Entypo').default;
    case 'font-awesome':
      return require('react-native-vector-icons/FontAwesome').default;
    case 'font-awesome-5':
      return require('react-native-vector-icons/FontAwesome5').default;
    case 'simple-line-icon':
      return require('react-native-vector-icons/SimpleLineIcons').default;
    case 'feather':
      return require('react-native-vector-icons/Feather').default;
    case 'antdesign':
    case 'ant-design':
      return require('react-native-vector-icons/AntDesign').default;
    case 'fontisto':
      return require('react-native-vector-icons/Fontisto').default;
    default:
      if (Object.prototype.hasOwnProperty.call(customIcons, type)) {
        return customIcons[type];
      }
      return require('react-native-vector-icons/MaterialIcons').default;
  }
};

export const getIconProperties = (node: React.ReactElement) => {
  const { type, name, size, color } = node.props;
  return {
    type,
    name,
    size,
    color,
  };
};
