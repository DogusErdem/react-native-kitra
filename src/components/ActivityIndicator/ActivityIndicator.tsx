import { StyleSheet, View, ActivityIndicator as RNActivityIndicator } from 'react-native';
import React from 'react';
import useComponentTheme from '../../core/hooks/useComponentTheme';
import type { ActivityIndicatorProps, FCCWD } from '../../types';
import { opacity } from '../../utilities';
import { applyDefaults } from '../../core/KitraProvider';

const ActivityIndicator:FCCWD<ActivityIndicatorProps> = ({ theme, children }) => {
  const { componentTheme } = useComponentTheme(theme, 'activityIndicator', 'default');
  return (
    <View
      testID="container"
      style={[styles.centeredView, { backgroundColor: opacity(String(componentTheme.default?.background), 80) }]}
    >
      <View style={[styles.modalView]}>
        {children || (
        <RNActivityIndicator
          size="large"
          color={componentTheme.default?.indicator}
        />
        ) }
      </View>
    </View>
  );
};

export default applyDefaults(ActivityIndicator);

const styles = StyleSheet.create({
  centeredView: {
    zIndex: 100,
    flex: 1,
    width: '100%',
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center',
  },
  modalView: {
    borderRadius: 20,
    width: 150,
    height: 150,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
