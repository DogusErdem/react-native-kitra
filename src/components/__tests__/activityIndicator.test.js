import { render } from '@testing-library/react-native';
import { Text } from 'react-native';
import { KitraProvider } from '../../core/KitraProvider';
import ActivityIndicator from '../ActivityIndicator/ActivityIndicator';
import { opacity } from '../../utilities';
/* eslint-disable no-undef */

const renderWithContext = children => render(<KitraProvider>{children}</KitraProvider>);

describe('ActiviyIndicator', () => {
  it('renders avtivity indicator with children', () => {
    // eslint-disable-next-line react/no-children-prop
    const tree = renderWithContext(<ActivityIndicator children={<Text>TEST</Text>} />);
    expect(tree).toMatchSnapshot();
  });
  it('renders avtivity indicator', () => {
    const tree = renderWithContext(<ActivityIndicator />);
    expect(tree).toMatchSnapshot();
  });
  it('renders with custom theme', () => {
    const tree = renderWithContext(
      <ActivityIndicator theme={{ default: { background: 'black', indicator: 'green' } }} />,
    );
    const container = tree.getByTestId('container');
    // const activityIndicator = tree.getByTestId('activityIndicator');

    expect(container).toHaveStyle({ backgroundColor: opacity('black', 80) });
    // expect(activityIndicator).toHaveStyle({ color: 'green' });
  });
});
