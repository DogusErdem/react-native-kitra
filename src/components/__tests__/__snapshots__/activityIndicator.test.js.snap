// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiviyIndicator renders avtivity indicator 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flex": 1,
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
          "zIndex": 100,
        },
        {
          "backgroundColor": "rgba(246,249,255,0.8)",
        },
      ]
    }
    testID="container"
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "borderRadius": 20,
            "height": 150,
            "justifyContent": "center",
            "width": 150,
          },
        ]
      }
    >
      <ActivityIndicator
        color="#195CEF"
        size="large"
      />
    </View>
  </View>
</View>
`;

exports[`ActiviyIndicator renders avtivity indicator with children 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flex": 1,
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
          "zIndex": 100,
        },
        {
          "backgroundColor": "rgba(246,249,255,0.8)",
        },
      ]
    }
    testID="container"
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "borderRadius": 20,
            "height": 150,
            "justifyContent": "center",
            "width": 150,
          },
        ]
      }
    >
      <Text>
        TEST
      </Text>
    </View>
  </View>
</View>
`;
